### PR TITLE
Add linear optimizer version that uses VMCBatch

### PR DESCRIPTION
**Pull request by markdewing**
##  Proposed changes

Convert linear optimizer to use VMCBatch for generating samples.  Similar to #2541.

Fixes from #2562 are included where appropriate.

## What type(s) of changes does this code introduce

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/DriverTraits.h [hunk 1]
```diff
@@ -18,7 +18,8 @@ enum class QMCRunType
   WF_TEST,
   VMC_BATCH,
   DMC_BATCH,
-  OPTIMIZE_BATCH
+  OPTIMIZE_BATCH,
```
Comment (ye-luo) [inline src/QMCDrivers/DriverTraits.h:21]: Is OPTIMIZE_BATCH actually used?

Comment (ye-luo) [reply src/QMCDrivers/DriverTraits.h:21]: I mean we use linear not opt. The same as we should only use linear_batch and there is no need to expose opt_batch even we need the class.

Comment (markdewing) [reply src/QMCDrivers/DriverTraits.h:21]: For now, it seems necessary to distinguish between the batched and non-batched versions.   After the non-batched versions get removed, this enum could be simplified.    But I'm not sure if I understood your comment correctly?

File: src/QMCDrivers/QMCDriverFactory.cpp [hunk 2]
```diff
@@ -92,7 +93,11 @@ QMCDriverFactory::DriverAssemblyState QMCDriverFactory::readSection(int curSerie
   if (curName != "qmc")
     qmc_mode = curName;
   int nchars = qmc_mode.size();
-  if (qmc_mode.find("linear") < nchars)
+  if (qmc_mode.find("linear_batch") < nchars)
+  {
+    das.new_run_type = QMCRunType::LINEAR_OPTIMIZE_BATCH;
```
Comment (ye-luo) [inline src/QMCDrivers/QMCDriverFactory.cpp:98]: I mean the enum is used to label drivers here. If we don't need "opt/opt_batch" but only "linear/linear_batch", no need to connect "opt/opt_batch". Anyway, should not block this PR.

**Pull request merged**